fleetspeak/src/e2etesting/lib/setup_components.go

```diff
@@ -0,0 +1,296 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	ptypes "github.com/golang/protobuf/ptypes"
+	duration "github.com/golang/protobuf/ptypes/duration"
+	daemonservicePb "github.com/google/fleetspeak/fleetspeak/src/client/daemonservice/proto/fleetspeak_daemonservice"
+	clientConfigPb "github.com/google/fleetspeak/fleetspeak/src/client/generic/proto/fleetspeak_client_generic"
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	spb "github.com/google/fleetspeak/fleetspeak/src/common/proto/fleetspeak"
+	cpb "github.com/google/fleetspeak/fleetspeak/src/config/proto/fleetspeak_config"
+	fcpb "github.com/google/fleetspeak/fleetspeak/src/server/components/proto/fleetspeak_components"
+	grpcServicePb "github.com/google/fleetspeak/fleetspeak/src/server/grpcservice/proto/fleetspeak_grpcservice"
+	servicesPb "github.com/google/fleetspeak/fleetspeak/src/server/proto/fleetspeak_server"
+	"google.golang.org/grpc"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"time"
+)
+
+var (
+	mysqlDatabase = flag.String("mysql_database", "", "MySQL database name to use")
+	mysqlUsername = flag.String("mysql_username", "", "MySQL username to use")
+	mysqlPassword = flag.String("mysql_password", "", "MySQL password to use")
+)
+
+type componentCmds struct {
+	serverCmd  *exec.Cmd
+	clientCmd  *exec.Cmd
+	serviceCmd *exec.Cmd
+}
+
+func (cc *componentCmds) killAll() {
+	if cc.serverCmd != nil && cc.serverCmd.Process != nil {
+		cc.serverCmd.Process.Kill()
+	}
+	if cc.clientCmd != nil && cc.clientCmd.Process != nil {
+		cc.clientCmd.Process.Kill()
+	}
+}
+
+// Runs a command and redirects its output to main stdout
+func runCommand(cmd *exec.Cmd) error {
+	stdoutIn, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderrIn, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	err = cmd.Start()
+	if err != nil {
+		return fmt.Errorf("cmd.Start() failed: %v", err)
+	}
+
+	// cmd.Wait() should be called only after we finish reading
+	// from stdoutIn and stderrIn.
+	// wg ensures that we finish
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		io.Copy(os.Stdout, stdoutIn)
+		wg.Done()
+	}()
+
+	_, err = io.Copy(os.Stderr, stderrIn)
+	if err != nil {
+		return fmt.Errorf("Copy streams error: %v", err)
+	}
+
+	wg.Wait()
+	cmd.Wait()
```

Comment:
It is important, that cmd.Wait() is called on the main thread. We're still calling it in a goroutine for the long running processes.

Wait() is a synchronization mechanism, which ensures that the child processes terminate, before the main thread terminates. When the child processes are not "detached" (continuing to run in the background), this causes less unexpected behavior.

Reply:
Yes, now it's clear to me. Will it be a good solution to call cmd.Wait() for server and client in killAll(), and call Wait() for service in componentCmds.start() just after we call `runCommand(cc.serviceCmd)`?

Reply:
> Yes, now it's clear to me. Will it be a good solution to call cmd.Wait() for server and client in killAll(),

Yes.

killAll() should kill the processes and call Wait(), blocking until they finish.

So we would have:
cc.clientCmd = ...
runCommand(cc.clientCmd)
cc.serverCmd = ...
runCommand(cc.serverCmd)
...
cc.killAll() // kills the processes and calls cmd.Wait() in this thread, blocking until done.

Meaning, runCommand() doesn't block.

> and call Wait() for service in componentCmds.start() just after we call `runCommand(cc.serviceCmd)`?

No, the respective Wait()s would be called in killAll()